### PR TITLE
Create a Node-RED stack

### LINKED ISSUE #427: Create a Node-RED stack
I would like to propose a Node-RED stack that can be used to build and deploy a node-red project using Appsody.

A draft PR for this will be forthcoming.

### PR DESCRIPTION
**Pull request by knolleary**
### Checklist:

- [x] Read the [Code of Conduct](https://github.com/appsody/website/blob/master/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/appsody/website/blob/master/CONTRIBUTING.md).

- [x] Followed the [commit message guidelines](https://github.com/appsody/website/blob/master/CONTRIBUTING.md#commit-message-guidelines).

- [x] Stack adheres to [Appsody stack structure](https://github.com/appsody/website/blob/master/content/docs/stacks/stack-structure.md) - as far as I can tell, yes, but part of the reason for raising this as a draft PR now is to get some more eyes on to it so we can see what I've missed.

### Contributing a new stack:

- Describe how application dependencies are managed:

A node-red project consists of a package.json file and the flow configuration file. Any user-defined dependencies are listed in the package.json file and installed as part of the stack build - this follows the same pattern as the existing nodejs stacks.

- Explain how Appsody file watcher is utilized:

The watcher watches the flow json file and rebuilds whenever it is changed. In the context of a Node-RED project, the flow json is the 'source code' the user will be changing.

- Describe other Appsody environment variables defined by the stack image:

None above and beyond what the sample/nodejs stacks provide.

- Describe any limitations and known issues:

Needs some more testing and sanity checking - hence why it is a (currently) a draft PR.

### Related Issues:

Closes #427

### CODE REVIEW

**Comment by sympatheticmoose** [comment]
@knolleary is this read for us to take a look at and review?

**Comment by knolleary** [comment]
@sympatheticmoose I think so... it all works from my perspective.

**Comment by Kamran64** [comment]
@knolleary In that case, can we take it off draft?

**Comment by knolleary** [comment]
@Kamran64 done

File: incubator/node-red/stack.yaml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+name: Node-RED
+version: 0.1
```
Comment (Kamran64) [inline incubator/node-red/stack.yaml]: Version field needs to be formatted major.minor.patch here - i.e. should be 0.1.0

**Comment by knolleary** [comment]
@Kamran64 thanks for the comments - all fixed.

Will look at the kube side of things as that's a space I've not properly explored with appsody yet.

File: incubator/node-red/templates/simple/package.json [hunk 2]
```diff
@@ -0,0 +1,12 @@
+{
+    "name": "SampleAppsody-Node-RED-Project",
+    "description": "A sample Node-RED Project created with Appsody",
+    "version": "0.0.1",
```
Comment (skoh7645) [inline incubator/node-red/templates/simple/package.json]: 0.0.1 --> 0.1.0

File: incubator/node-red/README.md [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+├── package.json
+└-- settings.js
+```
+
+The `package.json` file will have the structure:
+```
+{
+    "name": "AppsodyProject",
+    "description": "A Node-RED Project",
+    "version": "0.0.1",
```
Comment (skoh7645) [inline incubator/node-red/README.md]: 0.0.1 --> 0.1.0

**Comment by skoh7645** [comment]
I noticed you haven't included any `.js` files in the default template. We usually prefer to give the user a simple application in the template that users will get if they do an `init`.

**Comment by knolleary** [comment]
Hi @skoh7645 

this is a stack for Node-RED, not Node.js. So the default template provides a minimal Node-RED project - which consists of the package.json and flow.json files. For a simple project, there are no JavaScript files for the user to create.

File: incubator/node-red/templates/simple/package.json (hunk 2 above)
Comment (skoh7645) [inline incubator/node-red/templates/simple/package.json]: 0.0.1 --> 0.1.0

File: incubator/node-red/image/Dockerfile-stack [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+ENV APPSODY_RUN_ON_CHANGE="npm start"
+ENV APPSODY_RUN_KILL=true
+
+ENV APPSODY_DEBUG="npm run debug"
+ENV APPSODY_DEBUG_ON_CHANGE="npm run debug"
+ENV APPSODY_DEBUG_KILL=true
+
+ENV APPSODY_TEST="npm test"
+ENV APPSODY_TEST_ON_CHANGE=""
+ENV APPSODY_TEST_KILL=false
```
Comment (skoh7645) [inline incubator/node-red/image/Dockerfile-stack]: You can remove these environment variable as they are unused

**Comment by knolleary** [comment]
Thanks @skoh7645 - changes pushed to get the version number consistent throughout and remove those unused env vars.

File: incubator/node-red/image/Dockerfile-stack [hunk 5]
```diff
@@ -0,0 +1,30 @@
+FROM node:10
+
+ENV APPSODY_MOUNTS=/:/project/user-app
+ENV APPSODY_DEPS=/project/user-app/node_modules
+
+ENV APPSODY_WATCH_DIR=/project/user-app
+ENV APPSODY_WATCH_IGNORE_DIR=/project/user-app/node_modules
+ENV APPSODY_WATCH_REGEX="^.*.json$"
+
+ENV APPSODY_INSTALL="npm install --prefix user-app && npm audit fix --prefix user-app"
```
Comment (skoh7645) [inline incubator/node-red/image/Dockerfile-stack]: `APPSODY_INSTALL` --> `APPSODY_PREP` as `APPSODY_INSTALL` has been deprecated

**Comment by skoh7645** [comment]
Thanks @knolleary for making the requested changes. Have you considered adding a simple "hello world" (or Node-RED equivalent) application to the default template? The stack is also missing the `config/app-deploy.yaml` file Appsody uses for deployment that @Kamran64 has mentioned perviously.

> Hi @knolleary - Just took a quick look and noted a few things that jumped out at me.
> Looks good overall so far though!
> 
> Note: We now have the option to deploy to Kubernetes if there is a deployment file called `app-deploy.yaml` present in the `image/config` directory. You may want to take a look at how existing node stacks use this (If you choose to add it, that is). E.g. For [nodejs](https://github.com/appsody/stacks/blob/master/incubator/nodejs/image/config/app-deploy.yaml)

File: incubator/node-red/image/Dockerfile-stack (hunk 5 above)
Comment (knolleary) [reply incubator/node-red/image/Dockerfile-stack]: Where can I find out what the current valid env vars are? This stack is only a month old and I don't follow every change being made to know what has been deprecated and what hasnt.

**Comment by knolleary** [comment]
@skoh7645 yes, I do plan to update the default template with a sample app as well as add to the documentation. But that is follow-on work to getting the basic stack structure right which is my primary concern.

I'll look at adding the app-deploy file. I assume it can be much the same as the existing node.js one.

File: incubator/node-red/image/Dockerfile-stack (hunk 5 above)
Comment (skoh7645) [reply incubator/node-red/image/Dockerfile-stack]: Our stack env vars are documented here: https://appsody.dev/docs/stacks/environment-variables

Comment (knolleary) [reply incubator/node-red/image/Dockerfile-stack]: So I just need to rename the env var? It's purpose hasn't changed? I note the nodejs-express and nodejs stacks still use `APPSODY_INSTALL`.

Comment (skoh7645) [reply incubator/node-red/image/Dockerfile-stack]: Yes it is just renaming, the purpose has not changed. (Well spotted with the node stacks - we will try to remember to update the env var in the new stack releases!)

File: incubator/node-red/image/config/app-deploy.yaml [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+  # Add fields here
+  version: 1.0.0
+  applicationImage: APPSODY_DOCKER_IMAGE 
+  stack: APPSODY_STACK
+  service:
+    type: NodePort
+    port: APPSODY_PORT
+    annotations:
+      prometheus.io/scrape: 'true'
+  readinessProbe:
```
Comment (skoh7645) [inline incubator/node-red/image/config/app-deploy.yaml]: Can you also add the fields for service monitoring to this config file - our Appsody operator now supports creating service monitors CRs if there is a monitoring entry. See for [nodejs-express](https://github.com/appsody/stacks/blob/master/incubator/nodejs-express/image/config/app-deploy.yaml)

**Comment by knolleary** [comment]
I believe this PR has had all of the changes requested and can be re-reviewed - @Kamran64 @skoh7645

**Comment by knolleary** [comment]
Expanding the content of the simple application template is a follow-on piece of work. I don't plan to do that as part of this PR.

**Comment by skoh7645** [comment]
We decided to hold off merging this stack until the application is in as it would be concerning for the user, if the first thing they see when they hit that endpoint is "Not Found".

**Comment by knolleary** [comment]
Hi @skoh7645 - I've now added a default flow to the simple template.

**Comment by Kamran64** [comment]
LGTM - @skoh7645, thoughts?

File: incubator/node-red/image/Dockerfile-stack [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+ENV APPSODY_RUN_ON_CHANGE="npm start"
+ENV APPSODY_RUN_KILL=true
+
+ENV APPSODY_DEBUG="npm run debug"
+ENV APPSODY_DEBUG_ON_CHANGE="npm run debug"
+ENV APPSODY_DEBUG_KILL=true
+
+ENV APPSODY_TEST="npm test"
+
+COPY ./project /project
```
Comment (skoh7645) [inline incubator/node-red/image/Dockerfile-stack]: You also need to copy the `/config` folder into the container so `appsody deploy` has access to the deployment config

Comment (knolleary) [reply incubator/node-red/image/Dockerfile-stack]: Done

**Comment by skoh7645** [comment]
The image built with `appsody build` fails to run, we think it is a problem with Appmetrics 
Screenshot 2019-11-04 at 11 32 13

**Comment by sam-github** [comment]
Your stacktrace includes `/project/server.js:58`:
```
app.use(appSettings.httpAdminRoot,RED.httpAdmin);
```
That doesn't look like a middleware registration, and the error message would suggest that `ppSettings.httpAdminRoot` is `undefined`.

**Comment by knolleary** [comment]
Thanks for prompt @sam-github.

I had assumed this was an issue with appmetrics because I'd seen something similar with it in the past, but I've not had the time to dig into it.

But your comment made me spot the issue - you are right that `httpAdminRoot` is not defined, and by default it isn't expected to be. So this just needed a simple check before trying to mount the admin express app.

I've tested it locally and all is working now.

@skoh7645 thanks for your patience on this - been a busy few weeks.

File: incubator/node-red/README.md [hunk 8]
```diff
@@ -0,0 +1,86 @@
+# Node-RED Appsody Stack
+
+This is a stack for building a deployable Node-RED application using [Appsody](https://appsody.dev).
```
Comment (skoh7645) [inline incubator/node-red/README.md]: Could you mention here that `This stack is based on the `Node.js v10` runtime and allows you to build and deploy a new or existing Node-RED application using Appsody....

File: incubator/node-red/README.md [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+
+The `settings.js` file is a normal Node-RED settings file, but the stack will override
+some of the values:
+
+ - If running in development mode, `httpAdminRoot` is set to `/admin`.
+ - Otherwise, if `NODE_ENV` is set to `production`, the editor and Admin API will
+   be disabled entirely.
+
+
+Given this structure, the follow steps can be used to use the stack:
```
Comment (skoh7645) [inline incubator/node-red/README.md]: This should be under your `## Getting Started` section

File: incubator/node-red/README.md [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+The stack builds into the application a standard set of cloud-native endpoints
+to determine `readiness` and `liveness` of the application. These are exposed as:
+
+ - Readiness endpoint: http://localhost:3000/ready
+ - Liveness endpoint: http://localhost:3000/live
+ - Health check endpoint: http://localhost:3000/health
+
+## Templates
+
+This stack provides a single template - `simple`. It is the most minimal template
```
Comment (skoh7645) [inline incubator/node-red/README.md]: "the most minimal template" --> "a minimal template"

File: incubator/node-red/README.md [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+    "node-red": {
+        "settings": {
+            "flowFile": "flow.json",
+            "credentialsFile": "flow_cred.json"
+        }
+    }
+}
+```
+
+This stack will use the package.json file to install any additional dependencies
```
Comment (skoh7645) [inline incubator/node-red/README.md]: "package.json" --> "`package.json`"

File: incubator/node-red/README.md [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+the project has and also to identify the flow file - using the `node-red` section.
+
+If the package.json file doesn't have `node-red` section, the stack will use `flow.json`
```
Comment (skoh7645) [inline incubator/node-red/README.md]: "package.json" --> "`package.json`"

File: incubator/node-red/README.md (hunk 9 above)
Comment (skoh7645) [reply incubator/node-red/README.md]: "follow" --> "following"

**Pull request merged**